<a href="https://colab.research.google.com/github/nazia-noor/DATA690_Nazia/blob/main/assignment_06/Assignment_06_ipynb_txt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 06
### **Student name**- Nazia Noor
### **Course**- DATA690
### **Instructor**- Dr. Jay Wang


## Assignment 06
### Note:
- For visualization, you should use Plotly Express 
- Use use this notebook as your template and follow the instructions

The first half of this assignment is a repeat of assignment 05.

This gives you a chance to refresh.

The second half is new and gives you a chance to perform additional practices.

You also get a chance to use some of the Python libraries and techniques.

The links to the zip file is:

- https://data.ed.gov/dataset/college-scorecard-all-data-files-through-6-2020/resources (The page containing link to the zip file)

- https://ed-public-download.app.cloud.gov/downloads/CollegeScorecard_Raw_Data_01192021.zip (The link to download the zip file)

You can upload the zip file there or run `!wget` command to retrieve it directly, then run `!unzip` command to extract files.

You can also unzip and upload individual files to Colab. 

Your folder structure should look like this in your Colab enviroment:
```
- assignment_07.ipynb
- raw_data/
        - ...
        - 'MERGED1996_97_PP.csv',
        - 'MERGED2015_16_PP.csv',
        - ...
        - 'MERGED2017_18_PP.csv'
        - ...
```


In [ ]:
!wget https://ed-public-download.app.cloud.gov/downloads/CollegeScorecard_Raw_Data_01192021.zip

--2021-10-14 13:18:49--  https://ed-public-download.app.cloud.gov/downloads/CollegeScorecard_Raw_Data_01192021.zip
Resolving ed-public-download.app.cloud.gov (ed-public-download.app.cloud.gov)... 15.200.243.39, 15.205.92.235, 2600:1f12:18a:7d00:e2fd:78d3:a870:ea91, ...
Connecting to ed-public-download.app.cloud.gov (ed-public-download.app.cloud.gov)|15.200.243.39|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 385541173 (368M) [application/zip]
Saving to: ‘CollegeScorecard_Raw_Data_01192021.zip’

CollegeScorecard_Ra 100%[===================>] 367.68M  43.6MB/s    in 7.2s    

2021-10-14 13:18:57 (51.1 MB/s) - ‘CollegeScorecard_Raw_Data_01192021.zip’ saved [385541173/385541173]



In [ ]:
!unzip CollegeScorecard_Raw_Data_01192021.zip

Archive:  CollegeScorecard_Raw_Data_01192021.zip
   creating: Raw Data Files/
  inflating: Raw Data Files/MERGED2011_12_PP.csv  
  inflating: Raw Data Files/MERGED1999_00_PP.csv  
  inflating: Raw Data Files/MERGED2006_07_PP.csv  
  inflating: Raw Data Files/MERGED2014_15_PP.csv  
  inflating: Raw Data Files/Most-Recent-Cohorts-All-Data-Elements.csv  
  inflating: __MACOSX/Raw Data Files/._Most-Recent-Cohorts-All-Data-Elements.csv  
  inflating: Raw Data Files/MERGED2003_04_PP.csv  
  inflating: Raw Data Files/MERGED1996_97_PP.csv  
  inflating: Raw Data Files/FieldOfStudyData1516_1617_PP.csv  
  inflating: Raw Data Files/Most-Recent-Cohorts-Field-of-Study.csv  
  inflating: Raw Data Files/MERGED2012_13_PP.csv  
  inflating: Raw Data Files/MERGED2005_06_PP.csv  
  inflating: Raw Data Files/MERGED2009_10_PP.csv  
  inflating: Raw Data Files/FieldOfStudyData1617_1718_PP.csv  
  inflating: Raw Data Files/MERGED2018_19_PP.csv  
  inflating: Raw Data Files/MERGED2017_18_PP.csv  
  inflating

In [ ]:
# <1> 
# Import libraries

#(Write code here)
import os
import pandas as pd
import plotly.graph_objects as go

In [ ]:
# <2> 
# Display the current workin directory using os.getcwd()
# you would need to import a standard Python library called os which stands for operating system,
# so place that import statement in the previous cell.
# Since your notebook and your data files are not in # the same folder, 
# you want to make sure what is the current working folder and 
# how to access a data file in a different folder

#(Write code here)

x=os.getcwd ()
print (x)

/content


In [ ]:
! ls

 CollegeScorecard_Raw_Data_01192021.zip  'Raw Data Files'
 __MACOSX				  sample_data


In [ ]:
# <3> 
# os library has a method call listdir which generates a list of files in a directory/folder.
# use this method to assign the contents (list of file names) of the data folder to a variable and display it.
# # You can use ../ construct to traverse to the parent folder and then to another foloder parallel to the current folder

#(Write code here)
x=[file for file in os.listdir('/content/Raw Data Files')]
x  


['Crosswalks.zip',
 'MERGED2014_15_PP.csv',
 'MERGED2008_09_PP.csv',
 'MERGED2010_11_PP.csv',
 'MERGED2001_02_PP.csv',
 'MERGED2007_08_PP.csv',
 'MERGED2016_17_PP.csv',
 'MERGED1996_97_PP.csv',
 'Most-Recent-Cohorts-Field-of-Study.csv',
 'MERGED2009_10_PP.csv',
 'FieldOfStudyData1617_1718_PP.csv',
 'MERGED2004_05_PP.csv',
 'MERGED2011_12_PP.csv',
 'MERGED2015_16_PP.csv',
 'MERGED2000_01_PP.csv',
 'MERGED2006_07_PP.csv',
 'data.yaml',
 'MERGED1999_00_PP.csv',
 'MERGED2003_04_PP.csv',
 'MERGED2012_13_PP.csv',
 'FieldOfStudyData1516_1617_PP.csv',
 'FieldOfStudyData1415_1516_PP.csv',
 'MERGED1998_99_PP.csv',
 'MERGED2005_06_PP.csv',
 'MERGED2013_14_PP.csv',
 'MERGED2002_03_PP.csv',
 'MERGED2017_18_PP.csv',
 'MERGED1997_98_PP.csv',
 'MERGED2018_19_PP.csv',
 'Most-Recent-Cohorts-All-Data-Elements.csv']

In [ ]:
# <4> 
# The folder contains files that are not the yearly data files
# write code to remove the unwanted files from the list vsriable 
# Note: don't remove/delete these files from the folder in your drive
# For example, use the file extension to only keep the csv files 
# or use the name patter - data file name begins with "MERGED"
# You can use the concept of list comprehension to write just 
# one line of code as well as using a for loop, your choice.

#(Write code here)
y=["Raw Data Files/"+str(file) for file in x if file.startswith('MERGED')]
y


['Raw Data Files/MERGED2014_15_PP.csv',
 'Raw Data Files/MERGED2008_09_PP.csv',
 'Raw Data Files/MERGED2010_11_PP.csv',
 'Raw Data Files/MERGED2001_02_PP.csv',
 'Raw Data Files/MERGED2007_08_PP.csv',
 'Raw Data Files/MERGED2016_17_PP.csv',
 'Raw Data Files/MERGED1996_97_PP.csv',
 'Raw Data Files/MERGED2009_10_PP.csv',
 'Raw Data Files/MERGED2004_05_PP.csv',
 'Raw Data Files/MERGED2011_12_PP.csv',
 'Raw Data Files/MERGED2015_16_PP.csv',
 'Raw Data Files/MERGED2000_01_PP.csv',
 'Raw Data Files/MERGED2006_07_PP.csv',
 'Raw Data Files/MERGED1999_00_PP.csv',
 'Raw Data Files/MERGED2003_04_PP.csv',
 'Raw Data Files/MERGED2012_13_PP.csv',
 'Raw Data Files/MERGED1998_99_PP.csv',
 'Raw Data Files/MERGED2005_06_PP.csv',
 'Raw Data Files/MERGED2013_14_PP.csv',
 'Raw Data Files/MERGED2002_03_PP.csv',
 'Raw Data Files/MERGED2017_18_PP.csv',
 'Raw Data Files/MERGED1997_98_PP.csv',
 'Raw Data Files/MERGED2018_19_PP.csv']

In [ ]:
# <5> 
# Now that you have a clean list of the yearly files, you want to loop through them
# and read them into a dataframe one at a time. You only load six columns: 
# ["UNITID", "INSTNM", "STABBR", "REGION", "ADM_RATE", "TUITIONFEE_IN"].
# You should use "usecols" option of Pandas to avoid reading unwanted columns.
# You also want to add a new column call "YEAR" to differentiate the data frames from each other.
# The YEAR variable should be yyyy format so tht you can convert them into integer.
# if you use the format yyyy-yy (such as 1997-98 school year), you will not be able to convert them directly to integer
# if you use scatter plot, the YEAR needs to be converted to integer or float. 
# 
# You would use an empty list and append the yearly dataframes to the list.
# After all data files are loaded and appended to the list, you would use Pandas to concatenate them into a 
# new single data frame.
# Note: this exercise incorporates many techques we learned before
# - list (creating an empty, append an item to the list)
# - for loop 
# - read only the needed columns from a file (using usecols option)
# - add a new column to a data frame
# - concatenate multiple dataframes into a single one
# This exercise may appear a big challenging but it worths the effort. You will learn a lot and  love it. I promise.

#(Write code here)

def read_file(fname):
    df = pd.read_csv(fname,usecols=['UNITID', 'INSTNM', 'STABBR' , 'REGION' , 'ADM_RATE' , 'TUITIONFEE_IN'])
    df['year'] = fname.split('/')[1].split("_")[0].replace('MERGED','')
    return df  


In [ ]:
df_list = []

for fname in y:
  df_list.append(read_file(fname))


In [ ]:
#new datframe with "merged" file name
df=pd.concat(df_list)

In [ ]:
# <6> 
# explore the new dataframe (# of observations, varibles, head, tail, sample, missing values, statistics,etc.)

#(Write code here)

print("Number of observation",df.shape[0])
print("Number of variable",df.shape[1])


Number of observation 163331
Number of variable 7


In [ ]:
df.head()

UNITID                               INSTNM  ... TUITIONFEE_IN  year
0  100654             Alabama A & M University  ...        9096.0  2014
1  100663  University of Alabama at Birmingham  ...        7510.0  2014
2  100690                   Amridge University  ...        6900.0  2014
3  100706  University of Alabama in Huntsville  ...        9158.0  2014
4  100724             Alabama State University  ...        8720.0  2014

[5 rows x 7 columns]

In [ ]:
df.tail()

UNITID                                    INSTNM  ... TUITIONFEE_IN  year
6801  48511112         Georgia Military College - Dublin  ...        6120.0  2018
6802  48511113        Georgia Military College - Eastman  ...        6120.0  2018
6803  48616901  American College of Barbering - Florence  ...           NaN  2018
6804  49005401      HCI College - Fort Lauderdale Campus  ...       21420.0  2018
6805  49146401          ABC Adult School - Cabrillo Lane  ...           NaN  2018

[5 rows x 7 columns]

In [ ]:
df.sample(15)

UNITID                                INSTNM  ... TUITIONFEE_IN  year
744   130253               Sacred Heart University  ...       25300.0  2006
3464  214111   Montgomery County Community College  ...        3750.0  2011
911   135364        Luther Rice College & Seminary  ...        3000.0  2002
743   129215  Eastern Connecticut State University  ...        5045.0  2003
275   110459      Alliant International University  ...           NaN  2000
467   119368           Everest College-LA Wilshire  ...           NaN  2007
6584  489937     Piedmont International University  ...        9480.0  2017
1433  155618    Northwest Kansas Technical College  ...           NaN  2012
2439  179201          Sanford-Brown College-Fenton  ...       10300.0  2003
4090  212601                     Gannon University  ...           NaN  1997
1015  143233          Bethany Theological Seminary  ...           NaN  2013
2025  164517         AQUINAS COLLEGE-NEWTON CAMPUS  ...           NaN  1998
2243  174525           Oak Hills Christian College  ...       10180.0  2003
1323  146685        Lincoln Land Community College  ...           NaN  1998
2677  185828    New Jersey Institute of Technology  ...        7906.0  2002

[15 rows x 7 columns]

In [ ]:
df.isnull().sum()

UNITID                0
INSTNM                0
STABBR                0
REGION                1
ADM_RATE         118419
TUITIONFEE_IN     84136
year                  0
dtype: int64

In [ ]:
df.describe()

UNITID         REGION      ADM_RATE  TUITIONFEE_IN
count  1.633310e+05  163330.000000  44912.000000   79195.000000
mean   9.432395e+05       4.554693      0.692951   11665.636871
std    4.571721e+06       2.189217      0.217058    9986.003542
min    1.006360e+05       0.000000      0.000000       0.000000
25%    1.633380e+05       3.000000      0.555075    3776.000000
50%    2.127800e+05       5.000000      0.717500    9285.000000
75%    4.075680e+05       6.000000      0.857100   16148.000000
max    4.914640e+07       9.000000      1.000000   74787.000000

In [ ]:
# <7>
# the dataframe contains many years of data of all U.S. colleges.
# let's just look at UMBC
# filter/query the dataframe to retrieve only rows that belong to UMBC
# save the UMNC data to a new data frame. using a new variable 
# so that the old big data frame is still available for later use.

#(Write code here)

df_umbc=df.loc[df['INSTNM'].str.contains('University of Maryland-Baltimore County')]

df_umbc=df_umbc.sort_values(['year'])

df_umbc

UNITID                                   INSTNM  ... TUITIONFEE_IN  year
2151  163268  University of Maryland-Baltimore County  ...           NaN  1996
2049  163268  University of Maryland-Baltimore County  ...           NaN  1997
1982  163268  University of Maryland-Baltimore County  ...           NaN  1998
1955  163268  University of Maryland-Baltimore County  ...           NaN  1999
1926  163268  University of Maryland-Baltimore County  ...        5490.0  2000
1915  163268  University of Maryland-Baltimore County  ...        5910.0  2001
1893  163268  University of Maryland-Baltimore County  ...        6362.0  2002
1873  163268  University of Maryland-Baltimore County  ...        7388.0  2003
1830  163268  University of Maryland-Baltimore County  ...        8020.0  2004
1819  163268  University of Maryland-Baltimore County  ...        8520.0  2005
1799  163268  University of Maryland-Baltimore County  ...        8622.0  2006
1779  163268  University of Maryland-Baltimore County  ...        8708.0  2007
1767  163268  University of Maryland-Baltimore County  ...        8780.0  2008
1745  163268  University of Maryland-Baltimore County  ...        8872.0  2009
1710  163268  University of Maryland-Baltimore County  ...        9171.0  2010
1694  163268  University of Maryland-Baltimore County  ...        9467.0  2011
1660  163268  University of Maryland-Baltimore County  ...        9764.0  2012
1629  163268  University of Maryland-Baltimore County  ...       10068.0  2013
1610  163268  University of Maryland-Baltimore County  ...       10384.0  2014
1555  163268  University of Maryland-Baltimore County  ...       11006.0  2015
1510  163268  University of Maryland-Baltimore County  ...       11264.0  2016
1478  163268  University of Maryland-Baltimore County  ...       11518.0  2017
1415  163268  University of Maryland-Baltimore County  ...       11778.0  2018

[23 rows x 7 columns]

In [ ]:
# <8>
# Explore this UMBC dataframe (# of observations, varibles, head, tail, sample, missing values, statistics,etc.)

#(Write code here)

print("number of observation: ",df_umbc.shape[0])
print("number of variables: ",df_umbc.shape[1])

number of observation:  23
number of variables:  7


In [ ]:
df_umbc.head()

UNITID                                   INSTNM  ... TUITIONFEE_IN  year
2151  163268  University of Maryland-Baltimore County  ...           NaN  1996
2049  163268  University of Maryland-Baltimore County  ...           NaN  1997
1982  163268  University of Maryland-Baltimore County  ...           NaN  1998
1955  163268  University of Maryland-Baltimore County  ...           NaN  1999
1926  163268  University of Maryland-Baltimore County  ...        5490.0  2000

[5 rows x 7 columns]

In [ ]:
df_umbc.tail()

UNITID                                   INSTNM  ... TUITIONFEE_IN  year
1610  163268  University of Maryland-Baltimore County  ...       10384.0  2014
1555  163268  University of Maryland-Baltimore County  ...       11006.0  2015
1510  163268  University of Maryland-Baltimore County  ...       11264.0  2016
1478  163268  University of Maryland-Baltimore County  ...       11518.0  2017
1415  163268  University of Maryland-Baltimore County  ...       11778.0  2018

[5 rows x 7 columns]

In [ ]:
df_umbc.sample(10)

UNITID                                   INSTNM  ... TUITIONFEE_IN  year
2049  163268  University of Maryland-Baltimore County  ...           NaN  1997
1694  163268  University of Maryland-Baltimore County  ...        9467.0  2011
1893  163268  University of Maryland-Baltimore County  ...        6362.0  2002
1955  163268  University of Maryland-Baltimore County  ...           NaN  1999
1478  163268  University of Maryland-Baltimore County  ...       11518.0  2017
1926  163268  University of Maryland-Baltimore County  ...        5490.0  2000
2151  163268  University of Maryland-Baltimore County  ...           NaN  1996
1510  163268  University of Maryland-Baltimore County  ...       11264.0  2016
1779  163268  University of Maryland-Baltimore County  ...        8708.0  2007
1819  163268  University of Maryland-Baltimore County  ...        8520.0  2005

[10 rows x 7 columns]

In [ ]:
df_umbc.isnull().sum()

UNITID           0
INSTNM           0
STABBR           0
REGION           0
ADM_RATE         5
TUITIONFEE_IN    4
year             0
dtype: int64

In [ ]:
df_umbc.describe()

UNITID  REGION   ADM_RATE  TUITIONFEE_IN
count      23.0    23.0  18.000000      19.000000
mean   163268.0     2.0   0.640694    9004.842105
std         0.0     0.0   0.052728    1825.795421
min    163268.0     2.0   0.568300    5490.000000
25%    163268.0     2.0   0.597600    8270.000000
50%    163268.0     2.0   0.632100    8872.000000
75%    163268.0     2.0   0.689375   10226.000000
max    163268.0     2.0   0.717400   11778.000000

In [ ]:
# <9> 
# Plot UMBC's in-state tution overtime from 1996 to 2019 using Bar Chart

#(Write code here)

fig=go.Figure()

trace_UMBC= go.Bar(
    x= df_umbc["year"],
    y=df_umbc["TUITIONFEE_IN"]
    )

fig.add_trace(trace_UMBC)


fig.update_layout(title="UMBC tuition Fee Over The Years",
                  xaxis={"title":"Tuition Fee"},
                  yaxis={"title":"Year"}
                  )

In [ ]:
# <10> 
# Plot UMBC's in-state tution overtime from 1996 to 2019 using Line Chart

#(Write code here)

fig=go.Figure()

trace_UMBC= go.Line(
    x= df_umbc["year"],
    y=df_umbc["TUITIONFEE_IN"],
    mode= "markers+lines"
    )

fig.add_trace(trace_UMBC)


fig.update_layout(title="UMBC Tuition Fee Over The Years",
                  xaxis={"title":"Year"},
                  yaxis={"title":"Tuition Fee"}
                  )
fig.show()

/usr/local/lib/python3.7/dist-packages/plotly/graph_objs/_deprecations.py:385: DeprecationWarning:

plotly.graph_objs.Line is deprecated.
Please replace it with one of the following more specific types
  - plotly.graph_objs.scatter.Line
  - plotly.graph_objs.layout.shape.Line
  - etc.




In [ ]:
# <11>
# Now let's look at the tuition growth rate year over year
# We need to calculate UMBC tuition change percentage each year
# First convert the TUITIONFEE_IN column to a Python List

#(Write code here)

UMBC_tuition_fee_list=df_umbc["TUITIONFEE_IN"].to_list()
UMBC_tuition_fee_list


[nan,
 nan,
 nan,
 nan,
 5490.0,
 5910.0,
 6362.0,
 7388.0,
 8020.0,
 8520.0,
 8622.0,
 8708.0,
 8780.0,
 8872.0,
 9171.0,
 9467.0,
 9764.0,
 10068.0,
 10384.0,
 11006.0,
 11264.0,
 11518.0,
 11778.0]

In [ ]:
# Loop through the list and calculate the % change each year over the prior year
# This takes some effort. Not hard, just some abstract/logical thinking and some experiments
# Have fun on this one

 #(Write code here)

UMBC_PCT_Change=[]

for i in range(0,len(UMBC_tuition_fee_list)):
  pct_change=((UMBC_tuition_fee_list[i]-UMBC_tuition_fee_list[i-1])/UMBC_tuition_fee_list[i-1])*100
  UMBC_PCT_Change.append(pct_change)
  
print (UMBC_PCT_Change)

[nan, nan, nan, nan, nan, 7.650273224043716, 7.648054145516074, 16.12700408676517, 8.554412560909583, 6.234413965087282, 1.1971830985915493, 0.9974483878450475, 0.8268259072117593, 1.0478359908883828, 3.370153291253381, 3.227565151019518, 3.1372134783986483, 3.11347808275297, 3.1386571315057608, 5.989984591679507, 2.344175904052335, 2.254971590909091, 2.2573363431151243]


In [ ]:
# The resulting number has many decimal points which are unnecessary and not visually appealing.
# Round up the percentage to two decimal points
# You can use for loop. Or better, use list comprehension for simplicity/brevity

#(Write code here)
UMBC_PCT_Change1= [round(x,2) for x in UMBC_PCT_Change]
UMBC_PCT_Change1

[nan,
 nan,
 nan,
 nan,
 nan,
 7.65,
 7.65,
 16.13,
 8.55,
 6.23,
 1.2,
 1.0,
 0.83,
 1.05,
 3.37,
 3.23,
 3.14,
 3.11,
 3.14,
 5.99,
 2.34,
 2.25,
 2.26]

In [ ]:
# Add the list of the percentages to the dataframe as a new column ("PCT_CHANGE")
# Not as hard as you may think. If you get stuck, you think too hard. 
# Google it and you will find the answer.

#(Write code here)

df_umbc["Pct_change"]= UMBC_PCT_Change1
df_umbc

UNITID                                   INSTNM  ...  year  Pct_change
2151  163268  University of Maryland-Baltimore County  ...  1996         NaN
2049  163268  University of Maryland-Baltimore County  ...  1997         NaN
1982  163268  University of Maryland-Baltimore County  ...  1998         NaN
1955  163268  University of Maryland-Baltimore County  ...  1999         NaN
1926  163268  University of Maryland-Baltimore County  ...  2000         NaN
1915  163268  University of Maryland-Baltimore County  ...  2001        7.65
1893  163268  University of Maryland-Baltimore County  ...  2002        7.65
1873  163268  University of Maryland-Baltimore County  ...  2003       16.13
1830  163268  University of Maryland-Baltimore County  ...  2004        8.55
1819  163268  University of Maryland-Baltimore County  ...  2005        6.23
1799  163268  University of Maryland-Baltimore County  ...  2006        1.20
1779  163268  University of Maryland-Baltimore County  ...  2007        1.00
1767  163268  University of Maryland-Baltimore County  ...  2008        0.83
1745  163268  University of Maryland-Baltimore County  ...  2009        1.05
1710  163268  University of Maryland-Baltimore County  ...  2010        3.37
1694  163268  University of Maryland-Baltimore County  ...  2011        3.23
1660  163268  University of Maryland-Baltimore County  ...  2012        3.14
1629  163268  University of Maryland-Baltimore County  ...  2013        3.11
1610  163268  University of Maryland-Baltimore County  ...  2014        3.14
1555  163268  University of Maryland-Baltimore County  ...  2015        5.99
1510  163268  University of Maryland-Baltimore County  ...  2016        2.34
1478  163268  University of Maryland-Baltimore County  ...  2017        2.25
1415  163268  University of Maryland-Baltimore County  ...  2018        2.26

[23 rows x 8 columns]

In [ ]:
# Finaly, we can make the plot tuition growth rate year over year - bar first, then line chart

# (write code here)

fig=go.Figure()

trace_0= go.Bar(
    x= df_umbc["year"],
    y=df_umbc["Pct_change"]
    )

fig.add_trace(trace_0)


fig.update_layout(title="UMBC Tuition Fee Percentage Change Over The Years",
                  xaxis={"title":"Year"},
                  yaxis={"title":"Growth percentage"}
                  )

In [ ]:
fig=go.Figure()

trace_0= go.Line(
    x= df_umbc["year"],
    y=df_umbc["Pct_change"]
    )

fig.add_trace(trace_0)


fig.update_layout(title="UMBC Tuition Fee Percentage Change Over The Years",
                  xaxis={"title":"Year"},
                  yaxis={"title":"Growth percentage"}
                  )

/usr/local/lib/python3.7/dist-packages/plotly/graph_objs/_deprecations.py:385: DeprecationWarning:

plotly.graph_objs.Line is deprecated.
Please replace it with one of the following more specific types
  - plotly.graph_objs.scatter.Line
  - plotly.graph_objs.layout.shape.Line
  - etc.




In [ ]:
# <12>
# Since we want to do the same calculation for JHU. 
# Instead of doing it piecemeal as we did for UMBC
# Let's create a function which can be reused for any college.
# This function takes a list of tuitions and return a list of percentage changes year over year
# This function can be used later 

#(Write code here)


def percentage_change(uni_tution_fee_list):

  pct_change_list=[]

  for i in range(0,len(uni_tution_fee_list)):
    pct_change=((uni_tution_fee_list[i]-uni_tution_fee_list[i-1])/uni_tution_fee_list[i-1])*100
    pct_change_list.append(pct_change)
    pct_change1=[round(x,2) for x in pct_change_list]
    
  return pct_change1


In [ ]:
# <13>
# the dataframe contains many years of data of all U.S. colleges.
# let's just look at JHU
# filter/query the dataframe to retrieve only rows that belong to JHU
# save the JHU data to a new data frame. using a new variable 
# so that the old big data frame is still available for later use.

#(Write code here)
df_jhu=df.loc[df['INSTNM'].str.contains('Johns Hopkins University')]

df_jhu=df_jhu.sort_values(['year'])

df_jhu


UNITID                    INSTNM STABBR  ...  ADM_RATE  TUITIONFEE_IN  year
2139  162928  Johns Hopkins University     MD  ...       NaN            NaN  1996
2039  162928  Johns Hopkins University     MD  ...       NaN            NaN  1997
1973  162928  Johns Hopkins University     MD  ...       NaN            NaN  1998
1946  162928  Johns Hopkins University     MD  ...       NaN            NaN  1999
1917  162928  Johns Hopkins University     MD  ...       NaN        24930.0  2000
1906  162928  Johns Hopkins University     MD  ...    0.3431        26210.0  2001
1884  162928  Johns Hopkins University     MD  ...    0.3506        27390.0  2002
1865  162928  Johns Hopkins University     MD  ...    0.3060        28730.0  2003
1823  162928  Johns Hopkins University     MD  ...    0.3071        30140.0  2004
1812  162928  Johns Hopkins University     MD  ...    0.3509        31620.0  2005
1792  162928  Johns Hopkins University     MD  ...    0.2792        33900.0  2006
1772  162928  Johns Hopkins University     MD  ...    0.2571        35900.0  2007
1760  162928  Johns Hopkins University     MD  ...    0.2609        37700.0  2008
1738  162928  Johns Hopkins University     MD  ...    0.2782        39150.0  2009
1703  162928  Johns Hopkins University     MD  ...    0.2225        40680.0  2010
1687  162928  Johns Hopkins University     MD  ...    0.1940        42280.0  2011
1653  162928  Johns Hopkins University     MD  ...    0.1843        43390.0  2012
1622  162928  Johns Hopkins University     MD  ...    0.1813        45470.0  2013
1603  162928  Johns Hopkins University     MD  ...    0.1608        47060.0  2014
1548  162928  Johns Hopkins University     MD  ...    0.1412        48710.0  2015
1503  162928  Johns Hopkins University     MD  ...    0.1284        50410.0  2016
1473  162928  Johns Hopkins University     MD  ...    0.1254        52170.0  2017
1410  162928  Johns Hopkins University     MD  ...    0.1148        53740.0  2018

[23 rows x 7 columns]

In [ ]:
# <14> 
# Plot JHU's in-state tuition overtime from 1996 to 2019. 
# Let's plot bar and line chart together

#(Write code here)

fig=go.Figure()

trace_0=go.Bar(
    x=df_jhu["year"],
    y=df_jhu["TUITIONFEE_IN"]
    )

trace_1=go.Line(
    x=df_jhu["year"],
    y=df_jhu["TUITIONFEE_IN"]
)

fig.add_trace(trace_0)
fig.add_trace(trace_1)

fig.update_layout(title="JHU Tuition fee over the years",
                  xaxis={"title":"Year"},
                  yaxis={"title":"TUITIONFEE_IN"}
)

fig.show()

/usr/local/lib/python3.7/dist-packages/plotly/graph_objs/_deprecations.py:385: DeprecationWarning:

plotly.graph_objs.Line is deprecated.
Please replace it with one of the following more specific types
  - plotly.graph_objs.scatter.Line
  - plotly.graph_objs.layout.shape.Line
  - etc.




In [ ]:
# <15>
# Let's calculate the JHU tuition percentge growth using the function defined earlier

#(Write code here)

JHU_tuition_fee_list=df_jhu["TUITIONFEE_IN"].to_list()

JHU_PCT_Change=percentage_change(JHU_tuition_fee_list)
print (JHU_PCT_Change)

[nan, nan, nan, nan, nan, 5.13, 4.5, 4.89, 4.91, 4.91, 7.21, 5.9, 5.01, 3.85, 3.91, 3.93, 2.63, 4.79, 3.5, 3.51, 3.49, 3.49, 3.01]


In [ ]:
# <16>
# Add the list of percentage changes to the dataframe as a new column ("PCT_CHANGE")

#(Write code here)

df_jhu["Pct_change"]=JHU_PCT_Change
df_jhu

UNITID                    INSTNM STABBR  ...  TUITIONFEE_IN  year  Pct_change
2139  162928  Johns Hopkins University     MD  ...            NaN  1996         NaN
2039  162928  Johns Hopkins University     MD  ...            NaN  1997         NaN
1973  162928  Johns Hopkins University     MD  ...            NaN  1998         NaN
1946  162928  Johns Hopkins University     MD  ...            NaN  1999         NaN
1917  162928  Johns Hopkins University     MD  ...        24930.0  2000         NaN
1906  162928  Johns Hopkins University     MD  ...        26210.0  2001        5.13
1884  162928  Johns Hopkins University     MD  ...        27390.0  2002        4.50
1865  162928  Johns Hopkins University     MD  ...        28730.0  2003        4.89
1823  162928  Johns Hopkins University     MD  ...        30140.0  2004        4.91
1812  162928  Johns Hopkins University     MD  ...        31620.0  2005        4.91
1792  162928  Johns Hopkins University     MD  ...        33900.0  2006        7.21
1772  162928  Johns Hopkins University     MD  ...        35900.0  2007        5.90
1760  162928  Johns Hopkins University     MD  ...        37700.0  2008        5.01
1738  162928  Johns Hopkins University     MD  ...        39150.0  2009        3.85
1703  162928  Johns Hopkins University     MD  ...        40680.0  2010        3.91
1687  162928  Johns Hopkins University     MD  ...        42280.0  2011        3.93
1653  162928  Johns Hopkins University     MD  ...        43390.0  2012        2.63
1622  162928  Johns Hopkins University     MD  ...        45470.0  2013        4.79
1603  162928  Johns Hopkins University     MD  ...        47060.0  2014        3.50
1548  162928  Johns Hopkins University     MD  ...        48710.0  2015        3.51
1503  162928  Johns Hopkins University     MD  ...        50410.0  2016        3.49
1473  162928  Johns Hopkins University     MD  ...        52170.0  2017        3.49
1410  162928  Johns Hopkins University     MD  ...        53740.0  2018        3.01

[23 rows x 8 columns]

In [ ]:
# <17> Make a bar plot of the JHU percentage changes tuitions over time

#(Write code here)

fig=go.Figure()

trace_0= go.Bar(
    x= df_jhu["year"],
    y=df_jhu["Pct_change"]
    )

fig.add_trace(trace_0)

fig.update_layout(title="JHU Tuition Fee Percentage Change Over The Years",
                  xaxis={"title":"Year"},
                  yaxis={"title":"Growth percentage"}
                  )



In [ ]:
# <18>
# In order to plot both UMBC and JHU tuition change over time in the same plot,
# we need to combine the two datasets using the common key of YEAR 
# First make a umbc2 dataframe with only two columns needed. WE don't need other columns.
# Also change the column name from "PCT_CHANGE" to "UMBC_PCT" in preparation for the merge.
# This is because both umbc and jhu dataframe have the same column name "PCT_CHANGE",
# We rename them to there is no collision during the merge. BTW, Pandas handles collision gracefully,
# Feel free to try it without changing the column names.

#(Write code here)

df_umbc.rename(columns = {'Pct_change':'UMBC_PCT'}, inplace = True)

new_df_umbc=df_umbc[['year','UMBC_PCT']]



In [ ]:
# Then make a jhu2 dataframe with only two columns needed. WE don't need other columns.
# Also change the column name from "PCT_CHANGE" to "JHU_PCT"  in preparation for the merge

#(Write code here)

df_jhu.rename(columns = {'Pct_change':'JHU_PCT'}, inplace = True)

new_df_jhu=df_jhu[['year','JHU_PCT']]

In [ ]:
# Now, merger the two data frames
# and display the new dataframe

#(Write code here)


new_df=new_df_umbc.merge(new_df_jhu, on = 'year', how = 'left')
new_df

year  UMBC_PCT  JHU_PCT
0   1996       NaN      NaN
1   1997       NaN      NaN
2   1998       NaN      NaN
3   1999       NaN      NaN
4   2000       NaN      NaN
5   2001      7.65     5.13
6   2002      7.65     4.50
7   2003     16.13     4.89
8   2004      8.55     4.91
9   2005      6.23     4.91
10  2006      1.20     7.21
11  2007      1.00     5.90
12  2008      0.83     5.01
13  2009      1.05     3.85
14  2010      3.37     3.91
15  2011      3.23     3.93
16  2012      3.14     2.63
17  2013      3.11     4.79
18  2014      3.14     3.50
19  2015      5.99     3.51
20  2016      2.34     3.49
21  2017      2.25     3.49
22  2018      2.26     3.01

In [ ]:
# Plot UMBC and JHU separately on the same plot 
# However, this does not work since the bars overlap.

#(Write code here)

fig = go.Figure()

trace_UMBC = go.Line(
    x=new_df["year"],
    y=new_df["UMBC_PCT"],
    name="UMBC"
)

fig.add_trace(trace_UMBC)

trace_JHU = go.Line(
    x=new_df["year"],
    y=new_df["JHU_PCT"],
    name="JHU")

fig.add_trace(trace_JHU)

fig.update_layout(title="UMBC & JHU Tuition Fee Percentage Growth",
                  xaxis={"title":"Year"},
                  yaxis={"title":"PCT_Change"})


/usr/local/lib/python3.7/dist-packages/plotly/graph_objs/_deprecations.py:385: DeprecationWarning:

plotly.graph_objs.Line is deprecated.
Please replace it with one of the following more specific types
  - plotly.graph_objs.scatter.Line
  - plotly.graph_objs.layout.shape.Line
  - etc.




In [ ]:
# Plot them together using y=["UMBC_PCT","JHU_PCT"] option of bar plot

#(Write code here)

fig = go.Figure()
   
trace_UMBC=go.Bar(
    name="UMBC",
    x=new_df["year"],
    y=new_df["UMBC_PCT"]
    )
         
    
trace_JHU=go.Bar(
    name="JHU",
    x=new_df["year"],
    y=new_df["JHU_PCT"]
    )



fig.add_trace(trace_UMBC)
fig.add_trace(trace_JHU)

fig.update_layout(title="UMBC & JHU Tuition Fee Growth",
                  xaxis={"title":"Year"},
                  yaxis={"title":"PCT_Change"}
)


In [ ]:
# Plot them together using y=["UMBC_PCT","JHU_PCT"] option of line plot

#(Write code here)

fig = go.Figure()

    
trace_U=go.Line(
    name="UMBC",
    x=new_df["year"],
    y=new_df["UMBC_PCT"]
    )
         
    
trace_J=go.Line(
    name="JHU",
    x=new_df["year"],
    y=new_df["JHU_PCT"]
    )



fig.add_trace(trace_U)
fig.add_trace(trace_J)

fig.update_layout(title="UMBC vs JHU Tuition Fee Growth",
                  xaxis={"title":"Year"},
                  yaxis={"title":"PCT_Change"}
)


/usr/local/lib/python3.7/dist-packages/plotly/graph_objs/_deprecations.py:385: DeprecationWarning:

plotly.graph_objs.Line is deprecated.
Please replace it with one of the following more specific types
  - plotly.graph_objs.scatter.Line
  - plotly.graph_objs.layout.shape.Line
  - etc.




In [ ]:
# <19> 
# document your observation/conclusion of the comparison plots (bar/line)
# use the following Markdown cell

#### Based on my observation, from the above plot, UMBC's tuition fee percentage change rate fluctuated over the year. It rose sharply in 2003 but dropped dramatically after that. The highest percentage change rate was in 2003(16.13%) and the lowest percentage change rate was in 2008(.83%). 

#### JHU's tuition fee percentage rate was quite stable throughout the timeline with the highest percentage rate change in 2006 (7.21%) and lowest percentage rate in 2012(2.63%).



In [ ]:
# The end.